File: cooling_control.py
#!/usr/bin/env python3
"""
EAE Cooling Loop Control Logic
Author: Murray Kopit
Date: July 31, 2025

Controls coolant temperature for Inverter and DC-DC converter
"""

import time
import threading
from enum import Enum, auto
from dataclasses import dataclass
from typing import Optional

class SystemState(Enum):
    OFF = auto()
    INITIALIZING = auto()
    RUNNING = auto()
    ERROR = auto()
    EMERGENCY_STOP = auto()

@dataclass
class SensorData:
    temperature: float
    level_switch: bool
    ignition: bool

@dataclass
class ControlOutputs:
    pump_on: bool
    fan_on: bool
    fan_speed: int  # 0-100%

class CoolingController:
    def __init__(self):
        # Temperature thresholds (Celsius)
        self.TEMP_MIN = 50.0
        self.TEMP_TARGET = 65.0
        self.TEMP_MAX = 75.0
        self.TEMP_CRITICAL = 85.0
        
        # Fan control thresholds
        self.FAN_START_TEMP = 60.0
        self.FAN_MAX_TEMP = 80.0
        
        # System state
        self.state = SystemState.OFF
        self.sensors = SensorData(25.0, True, False)
        self.outputs = ControlOutputs(False, False, 0)
        
        # PID controller for fan speed
        self.pid = PIDController(kp=2.5, ki=0.5, kd=0.1, setpoint=self.TEMP_TARGET)
        
        # Safety timers
        self.pump_start_time = None
        self.low_level_time = None
        self.over_temp_time = None
        
        # Thread control
        self.running = False
        self.control_thread = None
        
    def start(self):
        """Start the control system"""
        self.running = True
        self.control_thread = threading.Thread(target=self._control_loop)
        self.control_thread.start()
        print("Cooling control system started")
        
    def stop(self):
        """Stop the control system"""
        self.running = False
        if self.control_thread:
            self.control_thread.join()
        self._shutdown_system()
        print("Cooling control system stopped")
        
    def update_sensors(self, temperature: float, level_switch: bool, ignition: bool):
        """Update sensor readings"""
        self.sensors.temperature = temperature
        self.sensors.level_switch = level_switch
        self.sensors.ignition = ignition
        
    def _control_loop(self):
        """Main control loop running at 10Hz"""
        while self.running:
            # State machine logic
            if self.state == SystemState.OFF:
                self._handle_off_state()
            elif self.state == SystemState.INITIALIZING:
                self._handle_init_state()
            elif self.state == SystemState.RUNNING:
                self._handle_running_state()
            elif self.state == SystemState.ERROR:
                self._handle_error_state()
            elif self.state == SystemState.EMERGENCY_STOP:
                self._handle_emergency_state()
                
            # Sleep for 100ms (10Hz update rate)
            time.sleep(0.1)
            
    def _handle_off_state(self):
        """Handle system in OFF state"""
        self.outputs.pump_on = False
        self.outputs.fan_on = False
        self.outputs.fan_speed = 0
        
        if self.sensors.ignition:
            print("Ignition ON - Starting initialization")
            self.state = SystemState.INITIALIZING
            
    def _handle_init_state(self):
        """Handle system initialization"""
        # Check coolant level
        if not self.sensors.level_switch:
            print("ERROR: Low coolant level detected")
            self.state = SystemState.ERROR
            return
            
        # Start pump
        self.outputs.pump_on = True
        self.pump_start_time = time.time()
        
        # Wait for circulation (2 seconds)
        if time.time() - self.pump_start_time > 2.0:
            print("Initialization complete - System running")
            self.state = SystemState.RUNNING
            
    def _handle_running_state(self):
        """Handle normal running state"""
        if not self.sensors.ignition:
            print("Ignition OFF - Shutting down")
            self.state = SystemState.OFF
            return
            
        # Safety checks
        if not self._perform_safety_checks():
            return
            
        # Temperature control
        self._control_temperature()
        
    def _handle_error_state(self):
        """Handle error conditions"""
        # Shut down pump and fan
        self.outputs.pump_on = False
        self.outputs.fan_on = False
        self.outputs.fan_speed = 0
        
        # Check if error condition cleared
        if self.sensors.level_switch and self.sensors.temperature < self.TEMP_MAX:
            if self.sensors.ignition:
                print("Error cleared - Restarting system")
                self.state = SystemState.INITIALIZING
            else:
                self.state = SystemState.OFF
                
    def _handle_emergency_state(self):
        """Handle emergency shutdown"""
        # Run fan at max speed even with pump off
        self.outputs.pump_on = False
        self.outputs.fan_on = True
        self.outputs.fan_speed = 100
        
        # Check if temperature reduced
        if self.sensors.temperature < self.TEMP_MAX:
            print("Temperature reduced - Attempting recovery")
            self.state = SystemState.ERROR
            
    def _perform_safety_checks(self) -> bool:
        """Perform safety checks, return False if unsafe"""
        # Check coolant level
        if not self.sensors.level_switch:
            if self.low_level_time is None:
                self.low_level_time = time.time()
            elif time.time() - self.low_level_time > 3.0:  # 3 second grace period
                print("ERROR: Coolant level low for >3 seconds")
                self.state = SystemState.ERROR
                return False
        else:
            self.low_level_time = None
            
        # Check critical temperature
        if self.sensors.temperature > self.TEMP_CRITICAL:
            print(f"CRITICAL: Temperature {self.sensors.temperature}°C exceeds limit")
            self.state = SystemState.EMERGENCY_STOP
            return False
            
        # Check over-temperature condition
        if self.sensors.temperature > self.TEMP_MAX:
            if self.over_temp_time is None:
                self.over_temp_time = time.time()
            elif time.time() - self.over_temp_time > 10.0:  # 10 second limit
                print("ERROR: Over-temperature for >10 seconds")
                self.state = SystemState.ERROR
                return False
        else:
            self.over_temp_time = None
            
        return True
        
    def _control_temperature(self):
        """Main temperature control logic"""
        temp = self.sensors.temperature
        
        # Pump control (always on when running)
        self.outputs.pump_on = True
        
        # Fan control with hysteresis
        if temp > self.FAN_START_TEMP:
            self.outputs.fan_on = True
            
            # PID control for fan speed
            self.outputs.fan_speed = self.pid.calculate(temp)
            
        elif temp < (self.FAN_START_TEMP - 5.0):  # 5°C hysteresis
            self.outputs.fan_on = False
            self.outputs.fan_speed = 0
            self.pid.reset()
            
    def _shutdown_system(self):
        """Safely shutdown all components"""
        self.outputs.pump_on = False
        self.outputs.fan_on = False
        self.outputs.fan_speed = 0
        self.state = SystemState.OFF
        

class PIDController:
    """Simple PID controller for fan speed control"""
    def __init__(self, kp: float, ki: float, kd: float, setpoint: float):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.setpoint = setpoint
        
        self.integral = 0.0
        self.last_error = 0.0
        self.last_time = time.time()
        
    def calculate(self, current_value: float) -> int:
        """Calculate PID output (0-100%)"""
        current_time = time.time()
        dt = current_time - self.last_time
        
        error = current_value - self.setpoint
        
        # Proportional term
        p_term = self.kp * error
        
        # Integral term with anti-windup
        self.integral += error * dt
        self.integral = max(-50, min(50, self.integral))  # Limit integral
        i_term = self.ki * self.integral
        
        # Derivative term
        d_term = self.kd * (error - self.last_error) / dt if dt > 0 else 0
        
        # Calculate output
        output = p_term + i_term + d_term
        
        # Update state
        self.last_error = error
        self.last_time = current_time
        
        # Clamp output to 0-100%
        return max(0, min(100, int(output)))
        
    def reset(self):
        """Reset controller state"""
        self.integral = 0.0
        self.last_error = 0.0
        self.last_time = time.time()


def main():
    """Demonstration of cooling control system"""
    controller = CoolingController()
    
    print("=== EAE Cooling Control System Demo ===")
    print("Simulating system operation...\n")
    
    # Start controller
    controller.start()
    
    try:
        # Simulate ignition on
        print("\n[t=0s] Turning ignition ON")
        controller.update_sensors(25.0, True, True)
        time.sleep(3)
        
        # Simulate temperature rise
        print("\n[t=3s] Temperature rising...")
        for temp in range(25, 70, 5):
            controller.update_sensors(float(temp), True, True)
            print(f"Temp: {temp}°C, Pump: {controller.outputs.pump_on}, "
                  f"Fan: {controller.outputs.fan_on}, Fan Speed: {controller.outputs.fan_speed}%")
            time.sleep(1)
            
        # Simulate steady state
        print("\n[t=12s] Steady state operation")
        controller.update_sensors(68.0, True, True)
        time.sleep(3)
        
        # Simulate low coolant level
        print("\n[t=15s] Simulating low coolant level")
        controller.update_sensors(68.0, False, True)
        time.sleep(5)
        
        # Restore coolant level
        print("\n[t=20s] Coolant level restored")
        controller.update_sensors(65.0, True, True)
        time.sleep(2)
        
        # Simulate over-temperature
        print("\n[t=22s] Simulating over-temperature condition")
        controller.update_sensors(88.0, True, True)
        time.sleep(2)
        
        # Cool down
        print("\n[t=24s] Cooling down")
        controller.update_sensors(70.0, True, True)
        time.sleep(2)
        
        # Ignition off
        print("\n[t=26s] Turning ignition OFF")
        controller.update_sensors(65.0, True, False)
        time.sleep(2)
        
    finally:
        controller.stop()
        print("\n=== Demo Complete ===")


if __name__ == "__main__":
    main()

File: cooling_control.cpp
/*
 * Copyright 2025 Murray Kopit
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/*
 * EAE Cooling Loop Control Logic - C++ Implementation
 * Author: Murray Kopit
 * Date: July 31, 2025
 *
 * Controls coolant temperature for Inverter and DC-DC converter
 */

#include <iostream>
#include <thread>
#include <chrono>
#include <atomic>
#include <cmath>
#include <iomanip>

enum class SystemState {
    OFF,
    INITIALIZING,
    RUNNING,
    ERROR,
    EMERGENCY_STOP
};

struct SensorData {
    float temperature;     // Celsius
    bool levelSwitch;      // True = OK, False = Low
    bool ignition;         // True = ON, False = OFF
};

struct ControlOutputs {
    bool pumpOn;
    bool fanOn;
    int fanSpeed;          // 0-100%
};

class PIDController {
private:
    float kp, ki, kd;
    float setpoint;
    float integral;
    float lastError;
    std::chrono::steady_clock::time_point lastTime;

public:
    PIDController(float p, float i, float d, float sp)
        : kp(p), ki(i), kd(d), setpoint(sp), integral(0.0f), lastError(0.0f) {
        lastTime = std::chrono::steady_clock::now();
    }

    int calculate(float currentValue) {
        auto now = std::chrono::steady_clock::now();
        float dt = std::chrono::duration<float>(now - lastTime).count();

        float error = currentValue - setpoint;

        // Proportional
        float pTerm = kp * error;

        // Integral with anti-windup
        integral += error * dt;
        if (integral > 50.0f) integral = 50.0f;
        if (integral < -50.0f) integral = -50.0f;
        float iTerm = ki * integral;

        // Derivative
        float dTerm = 0.0f;
        if (dt > 0) {
            dTerm = kd * (error - lastError) / dt;
        }

        // Update state
        lastError = error;
        lastTime = now;

        // Calculate output
        float output = pTerm + iTerm + dTerm;

        // Clamp to 0-100
        if (output < 0) output = 0;
        if (output > 100) output = 100;

        return static_cast<int>(output);
    }

    void reset() {
        integral = 0.0f;
        lastError = 0.0f;
        lastTime = std::chrono::steady_clock::now();
    }
};

class CoolingController {
private:
    // Temperature thresholds (Celsius)
    static constexpr float TEMP_MIN = 50.0f;
    static constexpr float TEMP_TARGET = 65.0f;
    static constexpr float TEMP_MAX = 75.0f;
    static constexpr float TEMP_CRITICAL = 85.0f;

    // Fan control thresholds
    static constexpr float FAN_START_TEMP = 60.0f;
    static constexpr float FAN_MAX_TEMP = 80.0f;
    static constexpr float FAN_HYSTERESIS = 5.0f;

    // Timing constants
    static constexpr float PUMP_INIT_TIME = 2.0f;      // seconds
    static constexpr float LOW_LEVEL_TIMEOUT = 3.0f;    // seconds
    static constexpr float OVER_TEMP_TIMEOUT = 10.0f;   // seconds

    // System state
    SystemState state;
    SensorData sensors;
    ControlOutputs outputs;

    // PID controller
    PIDController fanPID;

    // Timing
    std::chrono::steady_clock::time_point pumpStartTime;
    std::chrono::steady_clock::time_point lowLevelTime;
    std::chrono::steady_clock::time_point overTempTime;
    bool lowLevelTimerActive;
    bool overTempTimerActive;

    // Thread control
    std::atomic<bool> running;
    std::thread controlThread;

    void controlLoop() {
        // Use steady_clock for deterministic timing
        using clock = std::chrono::steady_clock;
        auto next = clock::now();
        const auto period = std::chrono::milliseconds(100);  // 10Hz = 100ms period

        while (running) {
            switch (state) {
                case SystemState::OFF:
                    handleOffState();
                    break;
                case SystemState::INITIALIZING:
                    handleInitState();
                    break;
                case SystemState::RUNNING:
                    handleRunningState();
                    break;
                case SystemState::ERROR:
                    handleErrorState();
                    break;
                case SystemState::EMERGENCY_STOP:
                    handleEmergencyState();
                    break;
            }

            // Calculate next wake time and sleep until then
            // This ensures consistent 100ms periods regardless of processing time
            next += period;
            std::this_thread::sleep_until(next);
        }
    }

    void handleOffState() {
        outputs.pumpOn = false;
        outputs.fanOn = false;
        outputs.fanSpeed = 0;

        if (sensors.ignition) {
            std::cout << "Ignition ON - Starting initialization" << std::endl;
            state = SystemState::INITIALIZING;
        }
    }

    void handleInitState() {
        // Check coolant level
        if (!sensors.levelSwitch) {
            std::cout << "ERROR: Low coolant level detected" << std::endl;
            state = SystemState::ERROR;
            return;
        }

        // Start pump
        if (!outputs.pumpOn) {
            outputs.pumpOn = true;
            pumpStartTime = std::chrono::steady_clock::now();
        }

        // Wait for circulation
        auto elapsed = std::chrono::steady_clock::now() - pumpStartTime;
        if (std::chrono::duration<float>(elapsed).count() > PUMP_INIT_TIME) {
            std::cout << "Initialization complete - System running" << std::endl;
            state = SystemState::RUNNING;
        }
    }

    void handleRunningState() {
        if (!sensors.ignition) {
            std::cout << "Ignition OFF - Shutting down" << std::endl;
            state = SystemState::OFF;
            return;
        }

        // Perform safety checks
        if (!performSafetyChecks()) {
            return;
        }

        // Temperature control
        controlTemperature();
    }

    void handleErrorState() {
        outputs.pumpOn = false;
        outputs.fanOn = false;
        outputs.fanSpeed = 0;

        // Check if error cleared
        if (sensors.levelSwitch && sensors.temperature < TEMP_MAX) {
            if (sensors.ignition) {
                std::cout << "Error cleared - Restarting system" << std::endl;
                state = SystemState::INITIALIZING;
            } else {
                state = SystemState::OFF;
            }
        }
    }

    void handleEmergencyState() {
        // Emergency cooling - run fan at max even with pump off
        outputs.pumpOn = false;
        outputs.fanOn = true;
        outputs.fanSpeed = 100;

        if (sensors.temperature < TEMP_MAX) {
            std::cout << "Temperature reduced - Attempting recovery" << std::endl;
            state = SystemState::ERROR;
        }
    }

    bool performSafetyChecks() {
        // Check coolant level
        if (!sensors.levelSwitch) {
            if (!lowLevelTimerActive) {
                lowLevelTime = std::chrono::steady_clock::now();
                lowLevelTimerActive = true;
            } else {
                auto elapsed = std::chrono::steady_clock::now() - lowLevelTime;
                if (std::chrono::duration<float>(elapsed).count() > LOW_LEVEL_TIMEOUT) {
                    std::cout << "ERROR: Coolant level low for >"
                             << LOW_LEVEL_TIMEOUT << " seconds" << std::endl;
                    state = SystemState::ERROR;
                    return false;
                }
            }
        } else {
            lowLevelTimerActive = false;
        }

        // Check critical temperature
        if (sensors.temperature > TEMP_CRITICAL) {
            std::cout << "CRITICAL: Temperature " << sensors.temperature
                     << "°C exceeds limit" << std::endl;
            state = SystemState::EMERGENCY_STOP;
            return false;
        }

        // Check over-temperature
        if (sensors.temperature > TEMP_MAX) {
            if (!overTempTimerActive) {
                overTempTime = std::chrono::steady_clock::now();
                overTempTimerActive = true;
            } else {
                auto elapsed = std::chrono::steady_clock::now() - overTempTime;
                if (std::chrono::duration<float>(elapsed).count() > OVER_TEMP_TIMEOUT) {
                    std::cout << "ERROR: Over-temperature for >"
                             << OVER_TEMP_TIMEOUT << " seconds" << std::endl;
                    state = SystemState::ERROR;
                    return false;
                }
            }
        } else {
            overTempTimerActive = false;
        }

        return true;
    }

    void controlTemperature() {
        float temp = sensors.temperature;

        // Pump always on when running
        outputs.pumpOn = true;

        // Fan control with hysteresis
        if (temp > FAN_START_TEMP) {
            outputs.fanOn = true;
            outputs.fanSpeed = fanPID.calculate(temp);
        } else if (temp < (FAN_START_TEMP - FAN_HYSTERESIS)) {
            outputs.fanOn = false;
            outputs.fanSpeed = 0;
            fanPID.reset();
        }
    }

public:
    CoolingController()
        : state(SystemState::OFF),
          fanPID(2.5f, 0.5f, 0.1f, TEMP_TARGET),
          lowLevelTimerActive(false),
          overTempTimerActive(false),
          running(false) {
        sensors = {25.0f, true, false};
        outputs = {false, false, 0};
    }

    void start() {
        running = true;
        controlThread = std::thread(&CoolingController::controlLoop, this);
        std::cout << "Cooling control system started" << std::endl;
    }

    void stop() {
        running = false;
        if (controlThread.joinable()) {
            controlThread.join();
        }
        shutdownSystem();
        std::cout << "Cooling control system stopped" << std::endl;
    }

    void updateSensors(float temperature, bool levelSwitch, bool ignition) {
        sensors.temperature = temperature;
        sensors.levelSwitch = levelSwitch;
        sensors.ignition = ignition;
    }

    void getStatus() const {
        std::cout << std::fixed << std::setprecision(1);
        std::cout << "Temp: " << sensors.temperature << "°C, "
                  << "Pump: " << (outputs.pumpOn ? "ON" : "OFF") << ", "
                  << "Fan: " << (outputs.fanOn ? "ON" : "OFF") << ", "
                  << "Fan Speed: " << outputs.fanSpeed << "%" << std::endl;
    }

private:
    void shutdownSystem() {
        outputs.pumpOn = false;
        outputs.fanOn = false;
        outputs.fanSpeed = 0;
        state = SystemState::OFF;
    }
};

// Demo program
int main() {
    std::cout << "=== EAE Cooling Control System Demo ===" << std::endl;
    std::cout << "Simulating system operation..." << std::endl << std::endl;

    CoolingController controller;
    controller.start();

    try {
        // Simulate ignition on
        std::cout << "\n[t=0s] Turning ignition ON" << std::endl;
        controller.updateSensors(25.0f, true, true);
        std::this_thread::sleep_for(std::chrono::seconds(3));

        // Simulate temperature rise
        std::cout << "\n[t=3s] Temperature rising..." << std::endl;
        for (int temp = 25; temp <= 70; temp += 5) {
            controller.updateSensors(static_cast<float>(temp), true, true);
            controller.getStatus();
            std::this_thread::sleep_for(std::chrono::seconds(1));
        }

        // Steady state
        std::cout << "\n[t=12s] Steady state operation" << std::endl;
        controller.updateSensors(68.0f, true, true);
        controller.getStatus();
        std::this_thread::sleep_for(std::chrono::seconds(3));

        // Low coolant
        std::cout << "\n[t=15s] Simulating low coolant level" << std::endl;
        controller.updateSensors(68.0f, false, true);
        std::this_thread::sleep_for(std::chrono::seconds(5));

        // Restore coolant
        std::cout << "\n[t=20s] Coolant level restored" << std::endl;
        controller.updateSensors(65.0f, true, true);
        std::this_thread::sleep_for(std::chrono::seconds(2));

        // Over-temperature
        std::cout << "\n[t=22s] Simulating over-temperature condition" << std::endl;
        controller.updateSensors(88.0f, true, true);
        controller.getStatus();
        std::this_thread::sleep_for(std::chrono::seconds(2));

        // Cool down
        std::cout << "\n[t=24s] Cooling down" << std::endl;
        controller.updateSensors(70.0f, true, true);
        controller.getStatus();
        std::this_thread::sleep_for(std::chrono::seconds(2));

        // Ignition off
        std::cout << "\n[t=26s] Turning ignition OFF" << std::endl;
        controller.updateSensors(65.0f, true, false);
        std::this_thread::sleep_for(std::chrono::seconds(2));

    } catch (...) {
        std::cerr << "Error in demo" << std::endl;
    }

    controller.stop();
    std::cout << "\n=== Demo Complete ===" << std::endl;

    return 0;
}
